lib/ret.ex

```diff
@@ -1,9 +1,207 @@
 defmodule Ret do
   @moduledoc """
-  Ret keeps the contexts that define your domain
-  and business logic.
+  The boundary of the Ret context.
+  """
+  alias Ret.{Account, Asset, Avatar, AvatarListing, OwnedFile, Project, Repo, Scene, SceneListing, Storage}
+  import Canada, only: [can?: 2]
+  import Ret.Schema, only: [is_serial_id: 1]
+  require Ecto.Query
+
+  @doc """
+  Obliterates the account with the given `account_id` and all its owned assets.
+
+  Returns `:ok` if successful, otherwise it returns `{:error, reason}`.
+
+  Error reasons are:
+
+  * `:forbidden` - the current user does not have sufficient authorization or
+    an account could not be found with the given `account_id`
+  * `:failed` - the account data was unable to be deleted for an unknown reason
 
-  Contexts are also responsible for managing your data, regardless
-  if it comes from the database, an external API or others.
   """
+  @spec delete_account(Account.id(), Account.t()) :: :ok | {:error, :forbidden | :failed}
+  def delete_account(account_id, %Account{} = current_user) when is_serial_id(account_id) do
+    account_to_delete = Repo.get(Account, account_id)
+
+    cond do
+      !account_to_delete ->
+        {:error, :forbidden}
+
+      not can?(current_user, delete_account(account_to_delete)) ->
+        {:error, :forbidden}
+
+      true ->
```

Comment:
Probably ok, but the fallthrough case being the actual deleting of accounts feels slightly scary to me. Would kind of like `can?(current_user, delete_account(account_to_delete)) ->` to be the case here with the fallthrough being forbidden. Unless this is more idiomatic.

Reply:
Fall-throughs for unexpected conditions are generally avoided in Erlang/Elixir. “Let it crash” is the mantra. With a `case` statement that means it’s best not to include a clause for a case you haven’t planned for. With `cond` that means the `true` condition isn’t a last-resort but instead the happy path. The conditions that yielded `{:error, reason}` can be thought of as [guard clauses](https://refactoring.com/catalog/replaceNestedConditionalWithGuardClauses.html), a fail-fast technique.